### PR TITLE
Unlock smart contracts deploy task

### PR DESCRIPTION
**Pull request by clemsos**
# Description

> In order to easily deploy to new networks we need a Hardhat task that can deploy a whole new environment for Unlock on any chain.

as specified in #7554

Also some cleanup to the contracts scripts - as discussed in #7436

```
npx hardhat deploy 
Starting deployments on localhost...
UNLOCK DEPLOYMENT > Deploying contracts on localhost with the account: 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266
UNLOCK DEPLOYMENT > isLocalNet : true
UNLOCK SETUP > Unlock (w proxy) deployed to: 0x9fE46736679d2D9a65F0992F2272dE9f3c7fa6e0 (tx: 0xcabe4a0649c7cc6c8130a11b970b0b476ab8db5f2ebc26a03641f9d43d96ad98)
PUBLIC LOCK > deployed to : 0xCf7Ed3AccA5a467e9e704C703E8D87F634fB0Fc9 (tx: 0x08284b2e9c80a8129c9cea321cb244c99ad6e619527df0e11ed46e001134af4f)
PUBLIC LOCK > Please verify it and call `npx hardhat set template` on the Unlock.
UNLOCK SETUP> Template set for Lock (tx: 0xc125e372992056a3e98224c13e9a73a16cd2991762c6178f82540e877af45354)
UDT SETUP > UDTv2 (w proxy) deployed to: 0x0165878A594ca255338adfa4d48449f69242Eb8F (tx: 0xe0aafb52a5ae64bfd1369054fcb9f91d91d041f1fc0798f886f11bcaafaa6a2f)
UNLOCK DEPLOYMENT > Pre-minted 1000000.0 UDT to deployer
UNLOCK DEPLOYMENT > grant minting permissions to the Unlock Contract
UNLOCK DEPLOYMENT > minter renounced minter role
WETH > deployed to : 0xa513E6E4b8f2a923D98304ec87F64353C4D5C853 (tx: 0x14aa3072aea528dd304515e5219ba1cb16af6fcdbf3d65a1969a3e1a633424a9
UNLOCK DEPLOYMENT > WETH deployed to : 0xa513E6E4b8f2a923D98304ec87F64353C4D5C853
UNISWAP/WETH SETUP > Using WETH contract at: 0xa513E6E4b8f2a923D98304ec87F64353C4D5C853
UNISWAP/WETH SETUP > Deploying Uniswap contracts using 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266
UNISWAP/WETH SETUP > Uniswap V2 Factory deployed to : 0x2279B7A0a67DB372996a5FaB50D91eAA73d2eBe6 (tx: 0xe24757f094b29bf969abc7f8ff243b500e71b48f4fdbfd505b3fb6688f989bad)
UNISWAP/WETH SETUP > Router V02 deployed to :  0x8A791620dd6260079BF849Dc5567aDC3F2FdC318 (tx: 0x4720f698af833dacb9833969b2c9badc60798790d7aac7291815fefc7907b91c)
UNLOCK DEPLOYMENT > UDT approved Uniswap Router for 1000.0 ETH
UNLOCK DEPLOYMENT > added liquidity to uniswap 1000.0
UNLOCK CONFIG > Unlock configured properly. (tx: 0x62c985c24653f54a30aff7f728a76921618a2adcfbde150eb040203bc3dc65f8)
UNISWAP ORACLE > Oracle deployed at: 0x0DCd1Bf9A1b36cE34237eEaFef220932846BCD82  (tx: 0xd337058402a3f8e871fc1ec23ffa7bcc996c5bf9b200dee0c46801b49ce37d53)
UNLOCK ORACLE CONFIG > Oracle configured properly. (tx: 0x0a2c05b5748b3964f0e944fb74eb025b2246e7e63af7c11265772616a60fcd8d)
```

# Checklist:

- [ ] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [ ] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [ ] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

**Comment by clemsos** [comment]
This is what is shown to user when executed
```
npx hardhat run scripts/deploy.js
UNLOCK SETUP > Deploying contracts on localhost with the account: 0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266
UNLOCK SETUP > isLocalNet : true
UNLOCK SETUP > Deployment info saved to ./deployments/localhost.
UNLOCK SETUP > Unlock proxy deployed to: 0x9fE46736679d2D9a65F0992F2272dE9f3c7fa6e0
UNLOCK SETUP > PublicLock deployed at 0xCf7Ed3AccA5a467e9e704C703E8D87F634fB0Fc9
UNLOCK SETUP > Template set for newly deployed lock
UNLOCK SETUP > UDT (v2) deployed to: 0x0165878A594ca255338adfa4d48449f69242Eb8F
UNLOCK SETUP > Pre-minted 20000 UDT
UNLOCK SETUP > grant minting permissions to the Unlock Contract
UNLOCK SETUP > minter renounced minter role
UNLOCK SETUP > unlock configured properly
```

File: smart-contracts/README.md [hunk 1]
```diff
@@ -113,10 +117,44 @@ npx hardhat verify --network polygon 0x7633dd082406861C384309c67576a4260C5775E0
...
+
+### Deploy all
+
+You can setup an entire test environments with all contracts using
+
+```
+npx hardhat deploy all
+```
+
+Or deploy different part separately
```
Comment (julien51) [inline smart-contracts/README.md:130]: Really like that granularity!

Comment (clemsos) [reply smart-contracts/README.md:130]: It makes the code slightly more verbose but also more usable

**Comment by clemsos** [comment]
still TODOs

- split big tasks into subtasks (cleaner) https://hardhat.org/guides/create-task.html#subtasks
- proper way to setup gnosis safe multisig owners
- figure out why oracle is crashing

File: smart-contracts/hardhat.config.js [hunk 2]
```diff
@@ -85,73 +81,15 @@ if (process.env.RUN_MAINNET_FORK) {
...
-    const publicLock = await PublicLock.deploy()
-
-    // eslint-disable-next-line no-console
-    console.log(
-      `New PublicLock template deployed at ${publicLock.address} on ${networkName} (${publicLock.deployTransaction.hash}). Please verify it and call setTemplate on the Unlock`
-    )
-  }
-)
+// tasks
+require('./tasks/accounts')
```
Comment (julien51) [inline smart-contracts/hardhat.config.js:85]: +1 for imports and a smaller file!

File: smart-contracts/hardhat.config.js [hunk 3]
```diff
@@ -166,6 +104,7 @@ module.exports = {
   },
   solidity: {
     compilers: [
+      { version: '0.4.18', settings },
```
Comment (julien51) [inline smart-contracts/hardhat.config.js:107]: Mind providing a bit of clarity about why this got added?

File: smart-contracts/scripts/deployments/index.js [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+      log(`WETH deployed to : ${wethAddress}`)
+    } else {
+      wethAddress = WETH[networkName]
+      log(`using WETH at: ${wethAddress}`)
+    }
+  }
+
+  // deploy uniswap v2 if needed
+  if (!uniswapFactoryAddress) {
+    const { router, factory } = await run('deploy:uniswap', { wethAddress })
```
Comment (julien51) [inline smart-contracts/scripts/deployments/index.js:101]: Oh that's cool! We can use it for Arbitrum on which only Uniswap v3 is deployed!

File: smart-contracts/hardhat.config.js (hunk 3 above)
Comment (clemsos) [reply smart-contracts/hardhat.config.js]: oh my bad... removed. I used it to compile WETH but found out the ABI was in `hardly-difficult-eth` package

**Pull request merged**